Changes for Heads on NovaCustom V560TU

mkopec opened the pull request:
- Fix building with FSP from dasharo-blobs
- Add PR0 lockdown in SMM
- Enable S3 on MTL
- Fixes for HAP on MTL
- Always pull FSP submodule, to ensure SplitFspBin is present
- Move IOE_P2SB below 4G to reduce MTRR fragmentation (faster boot on 64 and 96 GB RAM configs)
- Always put I2C and framebuffer below 4G when built for IA32, to make sure we can use them in coreboot 
- Fix for FSP GOP not initializing internal panel if an external display with res < 1920x1200 is connected
- JPEG patches for bootsplash, previously maintaned downstream in Heads

shuoliu0 committed:
soc/intel/xeon_sp: Support CHIPSET_LOCKDOWN_FSP

In a server platform many silicon specific register lock operations
are by default in FSP space. CHIPSET_LOCKDOWN_FSP provides an option
to make sure the codes could be used out-of-box to build products.

Change-Id: I8efcc1f27446be8e35f51e2568c4af6f8165486b
Signed-off-by: Shuo Liu <shuo.liu@intel.com>
Reviewed-on: https://review.coreboot.org/c/coreboot/+/82081
Reviewed-by: Angel Pons <th3fanbus@gmail.com>
Tested-by: build bot (Jenkins) <no-reply@coreboot.org>
Reviewed-by: Patrick Rudolph <patrick.rudolph@9elements.com>

nigeltao committed:
vc/wuffs: upgrade to Wuffs 0.4.0-alpha.8

We were previously at Wuffs 0.4.0-alpha.2. The C file was copied from
https://github.com/google/wuffs-mirror-release-c and its hash matches
https://github.com/google/wuffs-mirror-release-c/blob/90e4d81a6a8b7b601e8e568da32a105d7f7705e5/sync.txt#L9-L10

$ sha256sum src/vendorcode/wuffs/wuffs-v0.4.c
6c22caff4af929112601379a73f72461bc4719a5215366bcc90d599cbc442bb6  src/vendorcode/wuffs/wuffs-v0.4.c

Change-Id: Ie90d989384e0db2b23d7d1b3d9a57920ac8a95a2
Signed-off-by: Nigel Tao <nigeltao@golang.org>
Reviewed-on: https://review.coreboot.org/c/coreboot/+/83894
Reviewed-by: Felix Singer <service+coreboot-gerrit@felixsinger.de>
Reviewed-by: Nico Huber <nico.h@gmx.de>
Tested-by: build bot (Jenkins) <no-reply@coreboot.org>

nigeltao committed:
lib/jpeg: avoid calling malloc and free

Since commit 1d029b40c9de ("lib/jpeg: Replace decoder with Wuffs'
implementation"), a relatively large heap allocation is needed to decode
many JPEGs for use as work area. The prior decoder did not need this,
but also had many limitations in the JPEGs it could decode, was not as
memory-safe and quickly crashed under fuzzing.

This commit keeps using Wuffs' JPEG decoder, but it no longer requires
any heap allocation (and thus configuring the heap size depending on how
big a bootsplash image you want to support).

Change-Id: Ie4c52520cbce498539517c4898ff765365a6beba
Signed-off-by: Nigel Tao <nigeltao@golang.org>
Reviewed-on: https://review.coreboot.org/c/coreboot/+/83895
Tested-by: build bot (Jenkins) <no-reply@coreboot.org>
Reviewed-by: Nico Huber <nico.h@gmx.de>
Reviewed-by: Felix Singer <service+coreboot-gerrit@felixsinger.de>
Reviewed-by: Jonathon Hall <jonathon.hall@puri.sm>

mkopec committed:
vc/intel/fsp/fsp2_0/iot/meteorlake: add missing MemInfoHob from client vc dir

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

miczyg1 committed:
soc/intel/lockdown: Allow locking down SPI and LPC in SMM

Heads payload uses APM_CNT_FINALIZE SMI to set and lock down the SPI
controller with PR0 flash protection for pre-Skylake platforms.

Add new option to skip LPC and FAST SPI lock down in coreboot and move
it to APM_CNT_FINALIZE SMI handler. Reuse the INTEL_CHIPSET_LOCKDOWN
option to prevent issuing APM_CNT_FINALIZE SMI on normal boot path,
like it was done on pre-Skylake platforms. As the locking on modern
SOCs became more complicated, separate the SPI and LPC locking into
new modules to make linking to SMM easier.

The expected configuration to leverage the feautre is to unselect
INTEL_CHIPSET_LOCKDOWN and select SOC_INTEL_COMMON_SPI_LOCKDOWN_SMM.

Testing various microarchitectures happens on heads repository:
https://github.com/linuxboot/heads/pull/1818

TEST=Lock the SPI flash using APM_CNT_FINALIZE in heads on Alder Lake
(Protectli VP66xx) and Comet Lake (Protectli VP46xx) platforms. Check
if flash is unlocked in the heads recovery console. Check if flash is
locked in the kexec'ed OS.

Change-Id: Icbcc6fcde90e5b0a999aacb720e2e3dc2748c838
Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

mkopec committed:
clevo/mtl-h/ramstage.c: implement sleep mode option

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
soc/intel/mtl/fsp_params.c: wire up sleep mode option

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
soc/intel/cmn/blk/cse/cse.c: skip me_enable while ME is in Debug mode

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
Makefile.mk: Always pull FSP submodule

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
soc/intel/cmn/blk/graphics: ensure framebuffer is allocated <4G

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
soc/intel/cmn/blk/i2c: ensure resources are allocated <4G

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
mb/clevo/mtl-h: Set MMIO size to 2.5GB

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
soc/intel/mtl: move IOE P2SB BAR below 4G

This frees up two MTRRs for SPI flash caching

IOM_BASE_ADDR has to be hardcoded, otherwise iasl throws an error while
building

```
dsdt.asl   1374:     0x1600,,,)
Error    6050 -          ^ Length is not equal to fixed Min/Max window
```

It seems like IASL can't perform arithmetic operations inside
QWordMemory definitions.

See https://review.coreboot.org/c/coreboot/+/84216

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
mb/clevo/mtl-h/variants/igpu/data.vbt: Disable secondary display feature

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec commented:
Should go in before https://github.com/linuxboot/heads/pull/1846

miczyg1 commented on src/soc/intel/common/block/graphics/graphics.c:
```diff
@@ -278,6 +278,13 @@ static void graphics_dev_read_resources(struct device *dev)
+		/* Place framebuffer below 4G to allow drawing bootsplash */
+		struct resource *res_bar2 = find_resource(dev, PCI_BASE_ADDRESS_2);
+		res_bar2->limit = 0xffffffff;
+		res_bar2->flags &= ~(IORESOURCE_PCI64 | IORESOURCE_ABOVE_4G);
```
```suggestion
		res_bar2->flags &= ~IORESOURCE_ABOVE_4G;
```

Should be enough.  The BAR is still 64bit and we should keep that info. The resource limit set to UINT32_MAX  (0xffffffff) should make coreboot  allocate it under 4G. Please try it and apply in other places too.

mkopec replied:
Applied: https://github.com/Dasharo/coreboot/pull/609/commits/b890d38d0f14946e5f9c26631b6c64ac58e7d2ba

miczyg1 commented on src/soc/intel/meteorlake/Kconfig:
```diff
@@ -255,7 +255,7 @@ config PCR_BASE_ADDRESS
 config IOE_PCR_BASE_ADDRESS
 	hex
-	default 0x3fff0000000
+	default 0x60000000
```
Is it really necessary? That address  0x3fff0000000 will always be UC MMIO anyways and should not influence the MTRR allocation... I mean even if you have 8GB of memory or 96GB of memory, maximum possible address of RAM for 96GB would be 0x1800000000 (excluding any MMIO holes). If we add 2GB MMIO hole in 32bit space, then we would have 0x1880000000. Anything above it will have to be UC, always, no matter the IOE_P2SB BAR is in 32bit or 64bit space. Could you please try it?

mkopec replied:
Here's my cbmem without https://github.com/Dasharo/coreboot/pull/609/commits/5863a5f2118b963820c3ed69d8fdf9abc3f6e664 : [mtrr_96gb_before.log](https://github.com/user-attachments/files/18439047/mtrr_96gb_before.log)

With: [mtrr_96gb_after.log](https://github.com/user-attachments/files/18439048/mtrr_96gb_after.log)

I think the difference comes down to the last MTRR

miczyg1 replied:
The main difference is that the UC was the default type before and we struggled to find proper ranges for WB memory. Now that WB is the default (due to other changes) things are different. Still, we don't have to change this BAR even with 96GB of RAM.

mkopec replied:
but both have UC as default, though? In the very last MTRR dump before `MTRR check`

or am I misinterpreting the logs? @miczyg1

mkopec replied:
```
[DEBUG]  MTRR: default type WB/UC MTRR counts: 16/12.
[DEBUG]  MTRR: UC selected as default type.
[DEBUG]  MTRR: 0 base 0x0000000000000000 mask 0x00003fffc0000000 type 6
[DEBUG]  MTRR: 1 base 0x0000000040000000 mask 0x00003ffff8000000 type 6
[DEBUG]  MTRR: 2 base 0x0000000048000000 mask 0x00003ffffe000000 type 6
[DEBUG]  MTRR: 3 base 0x00000000f9800000 mask 0x00003fffff800000 type 5
[DEBUG]  MTRR: 4 base 0x00000000ff000000 mask 0x00003fffff000000 type 5
[DEBUG]  MTRR: 5 base 0x0000000100000000 mask 0x00003fff00000000 type 6
[DEBUG]  MTRR: 6 base 0x0000000200000000 mask 0x00003ffe00000000 type 6
[DEBUG]  MTRR: 7 base 0x0000000400000000 mask 0x00003ffc00000000 type 6
[DEBUG]  MTRR: 8 base 0x0000000800000000 mask 0x00003ff800000000 type 6
[DEBUG]  MTRR: 9 base 0x0000001000000000 mask 0x00003ff800000000 type 6
[ERROR]  Not enough MTRRs available! MTRR index is 10 with 10 MTRRs in total.
[ERROR]  Not enough MTRRs available! MTRR index is 11 with 10 MTRRs in total.
[WARN ]  Not enough MTRRs: 12 vs 10
[WARN ]  Unable to insert temporary MTRR range: 0x00000000f9800000 - 0x00000000f9ffffff size 0x00800000 type 5
```

vs

```
[DEBUG]  MTRR: default type WB/UC MTRR counts: 15/10.
[DEBUG]  MTRR: UC selected as default type.
[DEBUG]  MTRR: 0 base 0x0000000000000000 mask 0x00003fffc0000000 type 6
[DEBUG]  MTRR: 1 base 0x0000000040000000 mask 0x00003ffff8000000 type 6
[DEBUG]  MTRR: 2 base 0x0000000048000000 mask 0x00003ffffe000000 type 6
[DEBUG]  MTRR: 3 base 0x00000000f9800000 mask 0x00003fffff800000 type 5
[DEBUG]  MTRR: 4 base 0x00000000ff000000 mask 0x00003fffff000000 type 5
[DEBUG]  MTRR: 5 base 0x0000000100000000 mask 0x00003fff00000000 type 6
[DEBUG]  MTRR: 6 base 0x0000000200000000 mask 0x00003ffe00000000 type 6
[DEBUG]  MTRR: 7 base 0x0000000400000000 mask 0x00003ffc00000000 type 6
[DEBUG]  MTRR: 8 base 0x0000000800000000 mask 0x00003ff800000000 type 6
[DEBUG]  MTRR: 9 base 0x0000001000000000 mask 0x00003ff000000000 type 6
```

miczyg1 replied:
I am looking at the first occurrence, as is the the final one that will be applied( when temporary range gets removed - the one you see at the end). Before:

```
[DEBUG]  MTRR: Physical address space:
[DEBUG]  0x0000000000000000 - 0x000000000009ffff size 0x000a0000 type 6
[DEBUG]  0x00000000000a0000 - 0x00000000000bffff size 0x00020000 type 0
[DEBUG]  0x00000000000c0000 - 0x0000000049ffffff size 0x49f40000 type 6
[DEBUG]  0x000000004a000000 - 0x000000005fffffff size 0x16000000 type 0
[DEBUG]  0x0000000060000000 - 0x000000006fffffff size 0x10000000 type 1
[DEBUG]  0x0000000070000000 - 0x00000000ffffffff size 0x90000000 type 0
[DEBUG]  0x0000000100000000 - 0x000000189fffffff size 0x17a0000000 type 6
[DEBUG]  0x000003fff0000000 - 0x000003ffffffffff size 0x10000000 type 0
[DEBUG]  apic_id 0x20: MTRR: Fixed MSR 0x250 0x0606060606060606
[DEBUG]  apic_id 0x20: MTRR: Fixed MSR 0x258 0x0606060606060606
[DEBUG]  apic_id 0x20: MTRR: Fixed MSR 0x259 0x0000000000000000
[DEBUG]  apic_id 0x20: MTRR: Fixed MSR 0x268 0x0606060606060606
[DEBUG]  apic_id 0x20: MTRR: Fixed MSR 0x269 0x0606060606060606
[DEBUG]  apic_id 0x20: MTRR: Fixed MSR 0x26a 0x0606060606060606
[DEBUG]  apic_id 0x20: MTRR: Fixed MSR 0x26b 0x0606060606060606
[DEBUG]  apic_id 0x20: MTRR: Fixed MSR 0x26c 0x0606060606060606
[DEBUG]  apic_id 0x20: MTRR: Fixed MSR 0x26d 0x0606060606060606
[DEBUG]  apic_id 0x20: MTRR: Fixed MSR 0x26e 0x0606060606060606
[DEBUG]  apic_id 0x20: MTRR: Fixed MSR 0x26f 0x0606060606060606
[DEBUG]  apic_id 0x20 setup mtrr for CPU physical address size: 46 bits
[DEBUG]  MTRR: default type WB/UC MTRR counts: 7/11.
[DEBUG]  MTRR: WB selected as default type.
[DEBUG]  MTRR: 0 base 0x000000004a000000 mask 0x00003ffffe000000 type 0
[DEBUG]  MTRR: 1 base 0x000000004c000000 mask 0x00003ffffc000000 type 0
[DEBUG]  MTRR: 2 base 0x0000000050000000 mask 0x00003ffff0000000 type 0
[DEBUG]  MTRR: 3 base 0x0000000060000000 mask 0x00003ffff0000000 type 1
[DEBUG]  MTRR: 4 base 0x0000000070000000 mask 0x00003ffff0000000 type 0
[DEBUG]  MTRR: 5 base 0x0000000080000000 mask 0x00003fff80000000 type 0
[DEBUG]  MTRR: 6 base 0x000003fff0000000 mask 0x00003ffff0000000 type 0
```

After:

```
[DEBUG]  MTRR: Physical address space:
[DEBUG]  0x0000000000000000 - 0x000000000009ffff size 0x000a0000 type 6
[DEBUG]  0x00000000000a0000 - 0x00000000000bffff size 0x00020000 type 0
[DEBUG]  0x00000000000c0000 - 0x0000000049ffffff size 0x49f40000 type 6
[DEBUG]  0x000000004a000000 - 0x000000006fffffff size 0x26000000 type 0
[DEBUG]  0x0000000070000000 - 0x000000007fffffff size 0x10000000 type 1
[DEBUG]  0x0000000080000000 - 0x00000000ffffffff size 0x80000000 type 0
[DEBUG]  0x0000000100000000 - 0x000000189fffffff size 0x17a0000000 type 6
[DEBUG]  apic_id 0x20: MTRR: Fixed MSR 0x250 0x0606060606060606
[DEBUG]  apic_id 0x20: MTRR: Fixed MSR 0x258 0x0606060606060606
[DEBUG]  apic_id 0x20: MTRR: Fixed MSR 0x259 0x0000000000000000
[DEBUG]  apic_id 0x20: MTRR: Fixed MSR 0x268 0x0606060606060606
[DEBUG]  apic_id 0x20: MTRR: Fixed MSR 0x269 0x0606060606060606
[DEBUG]  apic_id 0x20: MTRR: Fixed MSR 0x26a 0x0606060606060606
[DEBUG]  apic_id 0x20: MTRR: Fixed MSR 0x26b 0x0606060606060606
[DEBUG]  apic_id 0x20: MTRR: Fixed MSR 0x26c 0x0606060606060606
[DEBUG]  apic_id 0x20: MTRR: Fixed MSR 0x26d 0x0606060606060606
[DEBUG]  apic_id 0x20: MTRR: Fixed MSR 0x26e 0x0606060606060606
[DEBUG]  apic_id 0x20: MTRR: Fixed MSR 0x26f 0x0606060606060606
[DEBUG]  apic_id 0x20 setup mtrr for CPU physical address size: 46 bits
[DEBUG]  MTRR: default type WB/UC MTRR counts: 6/9.
[DEBUG]  MTRR: WB selected as default type.
[DEBUG]  MTRR: 0 base 0x000000004a000000 mask 0x00003ffffe000000 type 0
[DEBUG]  MTRR: 1 base 0x000000004c000000 mask 0x00003ffffc000000 type 0
[DEBUG]  MTRR: 2 base 0x0000000050000000 mask 0x00003ffff0000000 type 0
[DEBUG]  MTRR: 3 base 0x0000000060000000 mask 0x00003ffff0000000 type 0
[DEBUG]  MTRR: 4 base 0x0000000070000000 mask 0x00003ffff0000000 type 1
[DEBUG]  MTRR: 5 base 0x0000000080000000 mask 0x00003fff80000000 type 0
```

But yeah, if we look at the last occurrence when temporary WP range is inserted, things fall apart for couple of reasons:

- there are two WP spaces (main flash and extended flash window)
- UC is stronger than WP, so UC MTRR can't cover WP range, thus it must fill the GAP between the two WP ranges with UC, contributing to more fragmented space

The FAST_SPI_SUPPORTS_EXT_BIOS_WINDOW should only be used when BIOS region is larger than 16MiB...

We already have a bunch of workarounds here. I don;t want to add more. Let's leave it as is

mkopec replied:
@miczyg1 I did test removing the extended BIOS window, it *still* was not enough with 96GB RAM:

```
[DEBUG]  MTRR: default type WB/UC MTRR counts: 13/11.
[DEBUG]  MTRR: UC selected as default type.
[DEBUG]  MTRR: 0 base 0x0000000000000000 mask 0x00003fffc0000000 type 6
[DEBUG]  MTRR: 1 base 0x0000000040000000 mask 0x00003ffff8000000 type 6
[DEBUG]  MTRR: 2 base 0x0000000048000000 mask 0x00003ffffe000000 type 6
[DEBUG]  MTRR: 3 base 0x00000000ff000000 mask 0x00003fffff000000 type 5
[DEBUG]  MTRR: 4 base 0x0000000100000000 mask 0x00003fff00000000 type 6
[DEBUG]  MTRR: 5 base 0x0000000200000000 mask 0x00003ffe00000000 type 6
[DEBUG]  MTRR: 6 base 0x0000000400000000 mask 0x00003ffc00000000 type 6
[DEBUG]  MTRR: 7 base 0x0000000800000000 mask 0x00003ff800000000 type 6
[DEBUG]  MTRR: 8 base 0x0000001000000000 mask 0x00003ff800000000 type 6
[DEBUG]  MTRR: 9 base 0x0000001800000000 mask 0x00003fff80000000 type 6
[ERROR]  Not enough MTRRs available! MTRR index is 10 with 10 MTRRs in total.
[WARN ]  Not enough MTRRs: 11 vs 10
[WARN ]  Unable to insert temporary MTRR range: 0x00000000ff000000 - 0x00000000ffffffff size 0x01000000 type 5
```
[mtrr_96gb_before_noextbios.log](https://github.com/user-attachments/files/18454139/mtrr_96gb_before_noextbios.log)

Maybe if I could align the top of lower usable RAM range to 1G it would've worked, but I could not get it to work using just FSP-M `MmioSize` param

miczyg1 commented on src/soc/intel/meteorlake/include/soc/iomap.h:
```diff
@@ -66,9 +66,10 @@
 #define IOE_P2SB_SIZE		(256 * MiB)
 
-#define IOM_BASE_ADDR		0x3fff0aa0000
+/* IOE_P2SB_BAR + 0xaa0000, but iasl refuses to perform arithmetics  */
```
What? Can you point to the faulting code?

mkopec replied:
Similar bug to https://review.coreboot.org/c/coreboot/+/84216

In my case setting this to `(IOE_P2SB_BAR + 0xaa0000)` results in this iasl error:

```
dsdt.asl   1374:     0x1600,,,)
Error    6050 -          ^ Length is not equal to fixed Min/Max window
```

coming from here: https://github.com/Dasharo/coreboot/blob/4e2fa994b412d5eafceac962d4e588354cf8c2f3/src/soc/intel/meteorlake/acpi/tcss.asl#L332-L335

macpijan replied:
Perhaps it's worth adding such references (if known) to the commit description

mkopec replied:
Right, amended: https://github.com/Dasharo/coreboot/pull/609/commits/d047598fdf4e9ce66ed26225d03ea9abf957f1d2

miczyg1 reviewed (changes requested):
While we are at limiting the necessary resources to 32bit, can you please limit it on the CSE HECI device as well? That way the allocation could work also for other builds than heads, where ME is active.

mkopec committed:
soc/intel/cmn/blk/cse/cse.c: ensure resources are allocated <4G

Without this patch, if RESOURCE_ALLOCATION_TOP_DOWN is enabled and
coreboot is compiled for IA32, the HECI device becomes inaccessible
after resource allocation.

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec commented:
> While we are at limiting the necessary resources to 32bit, can you please limit it on the CSE HECI device as well? That way the allocation could work also for other builds than heads, where ME is active.

Done https://github.com/Dasharo/coreboot/commit/5a313c76af2bc5e53e5c29d8c245b04f048ae776 (builds, but untested on a platform with ME enabled)

mkopec commented:
@miczyg1 can you take another look?
